### ISSUE TITLE
Replace coordinate columns with TARGET_RA, TARGET_DEC, (TARGET_)REF_EPOCH, PMRA, PMDEC to zpix table

### ISSUE THREAD

**Issue by stephjuneau** [desihub/specprod-db#6]
After reloading the `fiberassign` table ([Issue #5 ](https://github.com/desihub/specprod-db/issues/5)), the available coordinate columns TARGET_RA, TARGET_DEC, (TARGET_)REF_EPOCH, PMRA, PMDEC can be added to the `zpix` table, replacing the problematic `{MEAN|STD}_FIBER_{RA|DEC}` columns, which can be dropped at that point. 

@sbailey: is this a correct interpretation from what we discussed at this morning's data telecon?

**Comment by weaverba137** [comment]
In #4, PMRA, etc. were added to the fiberassign table. I believe the effort is now toward adding MEAN_FIBER_RA, etc. back to the zpix and ztile, so I suggest we repurpose this issue to represent that.

**Comment by weaverba137** [comment]
Note to self: assuming the VAC is ready soon, the load method could proceed as follows:

1. Truncate redshift and fiberassign tables.
2. Restore columns to redshift tables.
3. Start normal redshift loading stage.
4. During the pre-load hook, load the columns from the VAC and replace them.
5. Proceed with redshift loading.
6. Re-load fiberassign, so that bitmask corrections can be re-applied.

@stephjuneau, will the VAC(s) be row-matched to each row in ztile and zpix?

It may be possible to simplify the process further, but I wanted to get these notes written down.

**Comment by weaverba137** [comment]
@stephjuneau actually, let me rephrase: will the VACs be row matched to the zall *files*?  In principle the row ordering will be the same in the database, but for the above to work it has to be row-matched to the files.

**Comment by stephjuneau** [comment]
Hi @weaverba137, at the moment I sort by SURVEY, PROGRAM, TARGETID (in this order). However, I had to remove the rows with negative TARGETIDs because they fail a sanity check (unique TARGETIDs per SURVEY-PROGRAM combination). Do you want me to put those bad rows back in even though you skip them during the database ingest? I'm not sure at what step you subselect to only the subset of rows in the database that exclude negative TARGETIDs and SKY fibers. The database `zpix` table only has ~2e6 rows compared to ~2.8e6 in the zall-pix file.

**Comment by weaverba137** [comment]
After further discussion with @stephjuneau: the VACs contain a superset of the columns in the `zall` files, so the `zpix` and `ztile` tables could simply be reloaded from these VAC instead of the `zall` files.

Since the database excludes negative TARGETID anyway, it is not necessary to add these back in.

Testing to do: Double-check that columns that are not meant to be changed between `zall` and the VAC really are identical, including the data type.

**Comment by weaverba137** [comment]
I just noticed that the VACs also contain the bitmask corrections which are currently applied internally to the database.

I think this is OK, but are the corrections all the same as the values currently in `fuji_test`?'

Operationally, the bitmask correction stage of database loading will need to be made optional.

**Comment by stephjuneau** [comment]
Updates: the column datatypes (`dtype`) is the same for the VAC compared to the original zall-pix file. After talking with @sbailey , he mentioned that for the *new* columns, we might want to modify the dtype for the `*MJD` columns (I used float64 as this was the type for the `EXP_FIBERMAP` values but maybe it doesn't need to be so fine grained). The targeting bitmask corrections are the same as they come from a small 80KB patch file that I uploaded with the VAC and which itself comes from the `fuji` database query with the BITWISE_OR group_by compared to original zall-pix. This small patch file only includes TARGETID, SURVEY, PROGRAM and the list of 17 targeting columns. It is needed for `zpix` but not for `ztile`. For the workflow, you could probably still apply the bit mask correction as it would just re-apply the same correction and check before/after to confirm that indeed it was the same correction. It would still be useful for this to be optional in case we can fix it for Iron for instance.

**Comment by weaverba137** [comment]
@stephjuneau @sbailey Based on my recollections, I believe these are the further changes needed for the zpix and ztile tables:
```sql
ALTER TABLE fuji_test.zpix ADD COLUMN mean_fiber_ra REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN mean_fiber_dec REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN std_fiber_ra REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN std_fiber_dec REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN mean_delta_x REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN mean_delta_y REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN rms_delta_x REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN rms_delta_y REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN mean_psf_to_fiber_specflux REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN firstnight INTEGER NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN lastnight INTEGER NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN min_mjd DOUBLE PRECISION NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN mean_mjd DOUBLE PRECISION NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN max_mjd DOUBLE PRECISION NOT NULL;
--
--
--
ALTER TABLE fuji_test.ztile ADD COLUMN mean_fiber_ra REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_fiber_dec REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN std_fiber_ra REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN std_fiber_dec REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_delta_x REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_delta_y REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN rms_delta_x REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN rms_delta_y REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_psf_to_fiber_specflux REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN firstnight INTEGER NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN lastnight INTEGER NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN min_mjd DOUBLE PRECISION NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_mjd DOUBLE PRECISION NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN max_mjd DOUBLE PRECISION NOT NULL;
```
If that's not correct, that's fine, it hasn't been implemented yet, but this should give you time to review it.

**Comment by stephjuneau** [comment]
Hi @weaverba137 , with respect to this list above, the first thing I noticed is that for `ztile` there should already be an existing `lastnight` column so in my workflow I am not actually modifying this one. Also the fiber and delta columns were already there originally but then were dropped so do you still drop them and add them back in? Also another FYI for the columns that I'm recomputing, they are extended to include 'FIBER_X', 'FIBER_Y' for `ztile` because those columns are only coadded when working on single tiles. Here is what I use for the VAC:

(1) `zpix` VAC:
```
# Columns exist but values are updated
cols_to_replace = ['MEAN_DELTA_X','RMS_DELTA_X','MEAN_DELTA_Y','RMS_DELTA_Y','MEAN_FIBER_RA', \
                   'STD_FIBER_RA', 'MEAN_FIBER_DEC', 'STD_FIBER_DEC','MEAN_PSF_TO_FIBER_SPECFLUX']

# New columns
cols_to_add = ['MIN_MJD', 'MEAN_MJD', 'MAX_MJD', 'FIRSTNIGHT', 'LASTNIGHT']
```

(2) `ztile` VAC:
```
# Columns exist but values are updated
cols_to_replace = ['MEAN_FIBER_X', 'MEAN_FIBER_Y','MEAN_DELTA_X','RMS_DELTA_X','MEAN_DELTA_Y','RMS_DELTA_Y',\
                   'MEAN_FIBER_RA', 'STD_FIBER_RA', 'MEAN_FIBER_DEC', 'STD_FIBER_DEC','MEAN_PSF_TO_FIBER_SPECFLUX']

# LASTNIGHT already exists in the zall-tilecumulative file (leaving it as is)
cols_to_add = ['MIN_MJD', 'MEAN_MJD', 'MAX_MJD', 'FIRSTNIGHT'] 

```

**Comment by weaverba137** [comment]
I have to add back columns that were dropped last week. I'll check on LASTNIGHT in ztile when I get a chance.

**Comment by stephjuneau** [comment]
When you check on LASTNIGHT, could you please also check on MEAN_FIBER_X and MEAN_FIBER_Y in `ztile` ? 

(@sbailey  Interestingly, the coaddition.coadd_fibermap() function adds FIBER_X,Y to the MEAN columns but not to the STD columns when coadding over tiles -- while I wasn't sure if that was intentional or an omission, the VAC code behaves the same way for consistency)

**Comment by weaverba137** [comment]
OK, I agree that the original zall-tilecumulative-fuji.fits has `LASTNIGHT`, but the database does not have it.  Note however that the tile table does have it.  Is `LASTNIGHT` for `ztile` always the same as the value in tiles-fuji.fits?

**Comment by weaverba137** [comment]
There are no columns named 'mean*' in the `ztile` or `zpix` tables in `fuji_test` right now. It looks like `MEAN_FIBER_X` and `MEAN_FIBER_Y` were *never* added to either of those tables. Also, `zall-tilecumulative-fuji.fits` has `MEAN_FIBER_(X|Y)` but `zall-pix-fuji.fits` does *not*.  So it may have been overlooked.

**Comment by stephjuneau** [comment]
I would expect this to be the case with the exception of TILEID=80870 which has two values of LASTNIGHT in the zall-tile-cumulative file (email thread from today called "Double-checking on TILEID with two LASTNIGHT instances").

**Comment by weaverba137** [comment]
Revised column modifications:
```sql
ALTER TABLE fuji_test.zpix ADD COLUMN mean_fiber_ra REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN mean_fiber_dec REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN std_fiber_ra REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN std_fiber_dec REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN mean_delta_x REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN mean_delta_y REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN rms_delta_x REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN rms_delta_y REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN mean_psf_to_fiber_specflux REAL NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN firstnight INTEGER NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN lastnight INTEGER NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN min_mjd DOUBLE PRECISION NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN mean_mjd DOUBLE PRECISION NOT NULL;
ALTER TABLE fuji_test.zpix ADD COLUMN max_mjd DOUBLE PRECISION NOT NULL;
--
-- ztile has mean_fiber_(x|y) but zpix does not.
--
ALTER TABLE fuji_test.ztile ADD COLUMN mean_fiber_ra REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_fiber_dec REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN std_fiber_ra REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN std_fiber_dec REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_delta_x REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_delta_y REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN rms_delta_x REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN rms_delta_y REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_fiber_x REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_fiber_y REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_psf_to_fiber_specflux REAL NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN firstnight INTEGER NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN lastnight INTEGER NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN min_mjd DOUBLE PRECISION NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN mean_mjd DOUBLE PRECISION NOT NULL;
ALTER TABLE fuji_test.ztile ADD COLUMN max_mjd DOUBLE PRECISION NOT NULL;
```

**Comment by sbailey** [comment]
> (@sbailey Interestingly, the coaddition.coadd_fibermap() function adds FIBER_X,Y to the MEAN columns but not to the STD columns when coadding over tiles -- while I wasn't sure if that was intentional or an omission, the VAC code behaves the same way for consistency)

Yes intentional since the RMS_DELTA_X/Y columns contain the information about the scatter of the fiber positions about their intended location in the focal plane (although admittedly more obscure if you are use to looking for MEAN+STD columns and don't know to go looking for RMS_DELTA columns...)

Related for the record: note that pix files purposefully do not include a MEAN_FIBER_X/Y because the X/Y locations across different tiles can be very different and their mean is meaningless (pun sort of intended).  But pix files do keep the RMS_DELTA_X/Y because the scatter about the intended positions (even if they are different on each tile) is still meaningful.

**Comment by weaverba137** [comment]
@sbailey, I'm looking at the `zall-pix-fuji.fits` file, and it does have `MEAN_FIBER_(X|Y)`. I believe the individual `zpix` files all do as well.

**Comment by weaverba137** [comment]
Oops, I got confused. Hang on:

Corrected:
> I'm looking at the zall-pix-fuji.fits file, and it does have `MEAN_DELTA_(X|Y)`. I believe the individual zpix files all do as well.

In fact, zpix does not have `MEAN_FIBER_(X|Y)`.

The second draft column additions above should still be correct.
